core/src/main/java/org/springframework/ldap/core/SpringLdapClient.java

```diff
@@ -0,0 +1,666 @@
+/*
+ * Copyright 2002-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.ldap.core;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import javax.naming.Binding;
+import javax.naming.Name;
+import javax.naming.NameClassPair;
+import javax.naming.NameNotFoundException;
+import javax.naming.NamingEnumeration;
+import javax.naming.NamingException;
+import javax.naming.PartialResultException;
+import javax.naming.SizeLimitExceededException;
+import javax.naming.directory.Attributes;
+import javax.naming.directory.DirContext;
+import javax.naming.directory.ModificationItem;
+import javax.naming.directory.SearchControls;
+import javax.naming.directory.SearchResult;
+import javax.naming.ldap.LdapName;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.springframework.dao.EmptyResultDataAccessException;
+import org.springframework.dao.IncorrectResultSizeDataAccessException;
+import org.springframework.ldap.query.LdapQuery;
+import org.springframework.ldap.query.LdapQueryBuilder;
+import org.springframework.ldap.support.LdapUtils;
+
+/**
+ * Default implementation of {@link LdapClient}.
+ *
+ * @author Josh Cummings
+ * @since 3.0
+ */
+class SpringLdapClient implements LdapClient {
```

Comment:
It doesn't seem very consequential, but I wonder if for consistency with other projects, this should be `DefaultLdapClient` instead of Spring. I would expect a `SpringLdapClient` only if `LdapClient` weren't a Spring interface. Since it's hidden, it's mainly preference and for folks who like to click through and read source code in their IDEs.

Reply:
This goes back to a conversation with Rob long ago who prefers implementations don't start with the word "Default". Maybe a better name would be `ContextSourceLdapClient` since it uses a `ContextSource` instance.

Reply:
Since `WebClient`'s implementation is `DefaultWebClient`, I went ahead and changed this to be `DefaultLdapClient`.